### PR TITLE
SFMS "trigger" api

### PR DESCRIPTION
**Pull request by Sybrand**
- Endpoint for uploading SFMS output
- Script for uploading SFMS ouptut

# Test Links:
[Percentile Calculator](https://wps-pr-2228.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-2228.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-2228.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-2228.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-2228.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Fire Behaviour Advisory](https://wps-pr-2228.apps.silver.devops.gov.bc.ca/fire-behaviour-advisory)
[HFI Calculator](https://wps-pr-2228.apps.silver.devops.gov.bc.ca/hfi-calculator)
[FWI Calculator](https://wps-pr-2228.apps.silver.devops.gov.bc.ca/fwi-calculator)

### CODE REVIEW

**Review by conbrad — APPROVED** [review]
Should the API include the date for the actual processed?

**Review by conbrad — APPROVED** [review]

**Comment by Sybrand** [comment]
> Should the API include the date for the actual processed?

@conbrad I don't think there's a point putting the date on the API, since the processed date is always going to be assumed to be the date it was uploaded. The most the the uploader can do, is look at the create date of the file I supposed. We could add that - but realistically I think we should take the system date as the processed date.

File: api/app/routers/sfms.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+from app import config
+
+logger = logging.getLogger(__name__)
+
+router = APIRouter(
+    prefix="/sfms",
+)
+
+
+class FileLikeObject(io.IOBase):
```
Comment (conbrad) [inline api/app/routers/sfms.py:18]: Why this wrapper class? It's wrapping the exact same method signatures underneath?

**Review by conbrad — APPROVED** [review]

File: sfms/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# SFMS script
+
+## Overview
+
+This project contains a script that is intended to be run on the SFMS windows
+server. The script is intended to be run after SFMS geotiffs have been generated and uploads all geotiffs to the PSU API.
+
+The python script in this project is targeted at python 2.7
```
Comment (andrea-williams) [inline sfms/README.md:8]: Why py 2.7?

File: sfms/sfms.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            mimetype = mimetypes.guess_type(basename)[0] or 'application/octet-stream'
+        self.files.append((fieldname, basename, mimetype, body))
+        return
+
+    def get_binary(self):
+        """Return a binary buffer containing the form data, including attached files."""
+        part_boundary = '--' + self.boundary
+
+        binary = io.BytesIO()
+        needsCLRF = False
```
Comment (andrea-williams) [inline sfms/sfms.py:86]: what's the point of having this `needs CLRF` flag when it doesn't appear to ever change?

File: sfms/sfms.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    http.putheader('Content-length', str(len(form_buffer)))
+    http.endheaders()
+    http.send(form_buffer)
+    result = http.getresponse()
+
+    print('Status: ' + str(result.status))
+
+
+def main(ini):
+    """ Given some configuration filename, POST all files to the confured URL. """
```
Comment (andrea-williams) [inline sfms/sfms.py:145]:
```suggestion
    """ Given some configuration filename, POST all files to the configured URL. """
```

Comment (andrea-williams) [reply sfms/sfms.py:145]: Don't be making up words Sy.

**Review by andrea-williams — APPROVED** [review]
I think I'm happy with this? We shall see - dev chat pending.

File: api/app/routers/sfms.py (hunk 1 above)
Comment (Sybrand) [reply api/app/routers/sfms.py:18]: I've added a comment in the code: "The aiobotocore library expects a file-like object, but we can't pass the SpooledTemporaryFile
    object directly to aiobotocore. aiobotocore looks for a "tell" method, which isn't present
    on SpooledTemporaryFile. aiobotocore doesn't need an object with a tell method, and understands
    how to use IOBase, so we can wrap the SpooledTemporaryFile in a class that implements IOBase
    to make aiobotocore happy."

File: sfms/sfms.py (hunk 3 above)
Comment (Sybrand) [reply sfms/sfms.py:86]: I does change, one first iteration there is no \r\n - but on subsequent loops there is.

File: sfms/sfms.py (hunk 4 above)
Comment (Sybrand) [reply sfms/sfms.py:145]: I'm so good an conuring up words though!

File: api/app/routers/sfms.py (hunk 1 above)
Comment (conbrad) [reply api/app/routers/sfms.py:18]: Ok thanks, that's good to know. Do you still need to push up the commit with the comment?

File: api/app/utils/time.py [hunk 5]
```diff
@@ -1,4 +1,14 @@
...
+""" Functions for dealing with time.
+
+What's the deal with PDT,PST,UTC?????
+
+Best practice is to use UTC everywhere, but we have to deal with the fact that users think in
+terms of local time. Local time is confusing in BC, and especially so in wildfire because:
+- Solar noon is at 20h00 UTC.
+- Solar noon is at 12h00 PST. (So in the winter, the sun is overhead at noon)
+- Solar noon is at 13h00 PDT. (So in the summer, the sun is NOT overhead at noon)
+- That does noon even have to do with anything, when things like the daily FFMC is for 5pm.
```
Comment (andrea-williams) [inline api/app/utils/time.py:10]:
```suggestion
- What does noon even have to do with anything, when things like the daily FFMC is for 5pm.
```

File: sfms/README.md [hunk 6]
```diff
@@ -2,17 +2,38 @@
...
 ## Overview
 
-This project contains a script that is intended to be run on the SFMS windows
-server. The script is intended to be run after SFMS geotiffs have been generated and uploads all geotiffs to the PSU API.
+This project contains a script that is intended to be run on the SFMS windows server. The script is intended to be run after SFMS geotiffs have been generated and uploads all geotiffs to the PSU API.
 
-The python script in this project is targeted at python 2.7
+The python script in this project, `sfms.py` is targeted at python 2.7, since the SFMS server is running python 2.7.
+
+The pythong script, `sfms.py` does not require any external dependencies. [Python 2.7 is deprected](https://www.python.org/doc/sunset-python-2/) so we can't rely on pip being able to install any external dependencies.
```
Comment (andrea-williams) [inline sfms/README.md:9]:
```suggestion
The python script, `sfms.py` does not require any external dependencies. [Python 2.7 is deprected](https://www.python.org/doc/sunset-python-2/) so we can't rely on pip being able to install any external dependencies.
```

**Review by andrea-williams — APPROVED** [review]

**Pull request merged**